pkg/reconciler/configrevision_reconciler.go

```diff
@@ -0,0 +1,468 @@
+	for i := range revisions {
+		q := 0
+		if currentRevision != nil && revisions[i].Spec.Revision == currentRevision.Spec.Revision {
+			q = queued
+		}
+		revisions[i].Status.Queued = q
+		revisions[i].Status.Ongoing = cnt[revisions[i].Spec.Revision].ongoing
+		revisions[i].Status.Ready = cnt[revisions[i].Spec.Revision].ready
+		revisions[i].Status.Total = totalNodes
+		if err := crr.client.Status().Update(ctx, &revisions[i]); err != nil {
+			return fmt.Errorf("failed to update counters for revision %s: %w", revisions[i].Name, err)
+		}
+	}
+	return nil
+}
+
+func (crr *ConfigRevisionReconciler) revisionCleanup(ctx context.Context) error {
+	revisions, err := listRevisions(ctx, crr.client)
+	if err != nil {
+		return fmt.Errorf("error listing revisions: %w", err)
+	}
+
+	if len(revisions.Items) > 1 {
+		nodeConfigs, err := crr.listConfigs(ctx)
+		if err != nil {
+			return fmt.Errorf("error listing configs: %w", err)
+		}
+		if !revisions.Items[0].Status.IsInvalid && revisions.Items[0].Status.Ready == revisions.Items[0].Status.Total {
+			for i := 1; i < len(revisions.Items); i++ {
+				if countReferences(&revisions.Items[i], nodeConfigs.Items) == 0 {
+					if err := crr.client.Delete(ctx, &revisions.Items[i]); err != nil {
+						return fmt.Errorf("error deletring revision %s: %w", revisions.Items[i].Name, err)
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func countReferences(revision *v1alpha1.NetworkConfigRevision, configs []v1alpha1.NodeNetworkConfig) int {
+	refCnt := 0
+	for j := range configs {
+		if configs[j].Spec.Revision == revision.Spec.Revision {
+			refCnt++
+		}
+	}
+	return refCnt
+}
+
+func (crr *ConfigRevisionReconciler) listConfigs(ctx context.Context) (*v1alpha1.NodeNetworkConfigList, error) {
+	nodeConfigs := &v1alpha1.NodeNetworkConfigList{}
+	if err := crr.client.List(ctx, nodeConfigs); err != nil {
+		return nil, fmt.Errorf("error listing nodeConfigs: %w", err)
+	}
+	return nodeConfigs, nil
+}
+
+func (crr *ConfigRevisionReconciler) deployNodeConfig(ctx context.Context, node *corev1.Node, revision *v1alpha1.NetworkConfigRevision) error {
+	currentConfig := &v1alpha1.NodeNetworkConfig{}
+	if err := crr.client.Get(ctx, types.NamespacedName{Name: node.Name}, currentConfig); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("error getting NodeNetworkConfig object for node %s: %w", node.Name, err)
+		}
+		currentConfig = nil
+	}
+
+	if currentConfig != nil && currentConfig.Spec.Revision == revision.Spec.Revision {
+		// current config is the same as current revision - skip
+		return nil
+	}
+
+	newConfig, err := crr.createConfigForNode(node, revision)
+	if err != nil {
+		return fmt.Errorf("error preparing config for node %s: %w", node.Name, err)
+	}
+
+	if err := crr.deployConfig(ctx, newConfig, currentConfig, node); err != nil {
+		if errors.Is(err, InvalidConfigError) || errors.Is(err, context.DeadlineExceeded) {
```

Comment:
a) I think InvalidConfigError is not used anywhere
b) if the context DeadlineExceeded happens (e.g. operator can not communicate with K8s API) I would prefer a retry instead of marking the revision as invalid

Reply:
a) deleted InvalidConfigError
b) I've added explicitly 3 retries. If it won;t be able to connect within 3 tries, reconciler error will be reported.

Also now all actions like creation/update/deletion of objects etc. should be logged.